clang/test/BoundsSafety/CodeGen/array_subscript_agg.c

```diff
@@ -0,0 +1,1620 @@
+// LEGACY-NEXT:    [[TMP6:%.*]] = load i32, ptr [[IDX_ADDR]], align 4
+// LEGACY-NEXT:    [[IDXPROM:%.*]] = sext i32 [[TMP6]] to i64
+// LEGACY-NEXT:    [[ARRAYIDX:%.*]] = getelementptr [[STRUCT_FOO]], ptr [[WIDE_PTR_PTR]], i64 [[IDXPROM]]
+// LEGACY-NEXT:    call void @llvm.memcpy.p0.p0.i64(ptr align 4 [[RETVAL]], ptr align 4 [[ARRAYIDX]], i64 8, i1 false)
+// LEGACY-NEXT:    [[TMP7:%.*]] = load i64, ptr [[RETVAL]], align 4
+// LEGACY-NEXT:    ret i64 [[TMP7]]
+//
+struct Foo access_Foo_eb(struct Foo* __ended_by(end) start, int idx, struct Foo* end) {
+  return start[idx];
+}
+
+// NEW-LABEL: define dso_local i32 @access_Foo_member_bi(
+// NEW-SAME: ptr noundef [[PTR:%.*]], i32 noundef [[IDX:%.*]]) #[[ATTR0]] {
+// NEW-NEXT:  [[ENTRY:.*:]]
+// NEW-NEXT:    [[PTR_INDIRECT_ADDR:%.*]] = alloca ptr, align 8
+// NEW-NEXT:    [[IDX_ADDR:%.*]] = alloca i32, align 4
+// NEW-NEXT:    [[AGG_TEMP:%.*]] = alloca %"__bounds_safety::wide_ptr.bidi_indexable", align 8
+// NEW-NEXT:    store ptr [[PTR]], ptr [[PTR_INDIRECT_ADDR]], align 8
+// NEW-NEXT:    store i32 [[IDX]], ptr [[IDX_ADDR]], align 4
+// NEW-NEXT:    call void @llvm.memcpy.p0.p0.i64(ptr align 8 [[AGG_TEMP]], ptr align 8 [[PTR]], i64 24, i1 false)
+// NEW-NEXT:    [[WIDE_PTR_PTR_ADDR:%.*]] = getelementptr inbounds nuw %"__bounds_safety::wide_ptr.bidi_indexable", ptr [[AGG_TEMP]], i32 0, i32 0
+// NEW-NEXT:    [[WIDE_PTR_PTR:%.*]] = load ptr, ptr [[WIDE_PTR_PTR_ADDR]], align 8
+// NEW-NEXT:    [[TMP0:%.*]] = load i32, ptr [[IDX_ADDR]], align 4
+// NEW-NEXT:    [[IDXPROM:%.*]] = sext i32 [[TMP0]] to i64
+// NEW-NEXT:    [[ARRAYIDX:%.*]] = getelementptr [[STRUCT_FOO:%.*]], ptr [[WIDE_PTR_PTR]], i64 [[IDXPROM]]
+// NEW-NEXT:    [[WIDE_PTR_UB_ADDR:%.*]] = getelementptr inbounds nuw %"__bounds_safety::wide_ptr.bidi_indexable", ptr [[AGG_TEMP]], i32 0, i32 1
+// NEW-NEXT:    [[WIDE_PTR_UB:%.*]] = load ptr, ptr [[WIDE_PTR_UB_ADDR]], align 8
+// NEW-NEXT:    [[WIDE_PTR_LB_ADDR:%.*]] = getelementptr inbounds nuw %"__bounds_safety::wide_ptr.bidi_indexable", ptr [[AGG_TEMP]], i32 0, i32 2
+// NEW-NEXT:    [[WIDE_PTR_LB:%.*]] = load ptr, ptr [[WIDE_PTR_LB_ADDR]], align 8
+// NEW-NEXT:    [[TMP1:%.*]] = getelementptr [[STRUCT_FOO]], ptr [[ARRAYIDX]], i64 1, !annotation [[META2]]
+// NEW-NEXT:    [[TMP2:%.*]] = icmp ule ptr [[TMP1]], [[WIDE_PTR_UB]], !annotation [[META2]]
+// NEW-NEXT:    br i1 [[TMP2]], label %[[CONT:.*]], label %[[TRAP:.*]], !annotation [[META2]]
+// NEW:       [[TRAP]]:
+// NEW-NEXT:    call void @llvm.ubsantrap(i8 25) #[[ATTR3]], !annotation [[META2]]
+// NEW-NEXT:    unreachable, !annotation [[META2]]
+// NEW:       [[CONT]]:
+// NEW-NEXT:    [[TMP3:%.*]] = icmp ule ptr [[ARRAYIDX]], [[TMP1]], !annotation [[META2]]
+// NEW-NEXT:    br i1 [[TMP3]], label %[[CONT2:.*]], label %[[TRAP1:.*]], !annotation [[META2]]
+// NEW:       [[TRAP1]]:
+// NEW-NEXT:    call void @llvm.ubsantrap(i8 25) #[[ATTR3]], !annotation [[META2]]
```

Comment:
I wondered why this is different from legacy and it looks like this is the combined effect of both elem_size + array_subscript_agg. I initially thought we'd only compare the diff of array_subscript_agg, but I think this is fine too.

Reply:
@rapidsna The reason this test is doing both `access_size` and `array_subscript_agg` together is because in all practical cases we will use these new checks together but you're right that is does make the comparision between `LEGACY` and `NEW` a little confusing here.